### PR TITLE
Miscellaneous minor refactorings on the FSharp API

### PR DESCRIPTION
**Pull request by neoeinstein**
I figure a directed pull request to you would be better since I don't expect this to actually be merged as is. It does have some of my thoughts on rewriting some of the API internals and reordering things to take advantage of the functional composition. Let me know what you think, what you do/don't like, and I can fix this up a bit.

Also, I was wondering a bit why akka.net uses such heavy type annotations in the definitions. As part of the clean up, I think that we could benefit from removing those that the F# compiler is able to easily imply.

### CODE REVIEW

File: src/core/Akka.FSharp/Spawning.fs [hunk 1]
```diff
@@ -12,129 +12,135 @@ open Microsoft.FSharp.Quotations
...
+    type SpawnOption =
         | Deploy of Deploy
         | Router of Akka.Routing.RouterConfig
         | SupervisorStrategy of SupervisorStrategy
         | Dispatcher of string
         | Mailbox of string
-    
-    let rec applySpawnOptions (props : Props) (opt : SpawnOption list) : Props = 
+
+    let rec applySpawnOptions opt (props : Props) =
```
Comment (Horusiath) [inline src/core/Akka.FSharp/Spawning.fs:41]: While this is not a big thing, it's still breaking the existing API.

File: src/core/Akka.FSharp/Spawning.fs [hunk 2]
```diff
@@ -12,129 +12,135 @@ open Microsoft.FSharp.Quotations
...
-    /// <param name="expr">F# expression compiled down to receive function used by actor for response for incoming request</param>
+    /// <param name="actorExpr">F# expression compiled down to receive function used by actor for response for incoming request</param>
     /// <param name="options">List of options used to configure actor creation</param>
-    let spawne (actorFactory : IActorRefFactory) (name : string) 
-        (expr : Expr<Actor<'Message> -> Cont<'Message, 'Returned>>) (options : SpawnOption list) : ActorRef<'Message> = 
-        let e = Linq.Expression.ToExpression(fun () -> new FunActor<'Message, 'Returned>(expr))
-        let props = applySpawnOptions (Props.Create e) options
-        typed (actorFactory.ActorOf(props, name)) :> ActorRef<'Message>
-    
+    let spawne actorFactory options name (actorExpr : Expr<Actor<'Message> -> Cont<'Message, 'Returned>>) : ActorRef<'Message> =
```
Comment (Horusiath) [inline src/core/Akka.FSharp/Spawning.fs:72]: The Akka.FSharp convention, we follow, is to explicitly type each parameter used in public functions/methods. This way we have better control over incoming changes and reduce the risk of accidental breaking the API.

File: src/core/Akka.FSharp/Spawning.fs [hunk 3]
```diff
@@ -12,129 +12,135 @@ open Microsoft.FSharp.Quotations
...
-    /// <param name="f">Used by actor for handling response for incoming request</param>
+    /// <param name="actor">Used by actor for handling response for incoming request</param>
     /// <param name="options">List of options used to configure actor creation</param>
-    let spawnOpt (actorFactory : IActorRefFactory) (name : string) (f : Actor<'Message> -> Cont<'Message, 'Returned>) 
-        (options : SpawnOption list) : ActorRef<'Message> = 
-        let e = Linq.Expression.ToExpression(fun () -> new FunActor<'Message, 'Returned>(f))
-        let props = applySpawnOptions (Props.Create e) options
-        typed (actorFactory.ActorOf(props, name)) :> ActorRef<'Message>
-    
+    let spawnOpt actorFactory options name (actor : Actor<'Message> -> Cont<'Message, 'Returned>) : ActorRef<'Message> =
```
Comment (Horusiath) [inline src/core/Akka.FSharp/Spawning.fs:85]: As I've said this is (seriously) breaking API change, while it's possible advantage is subject to discussion.

**Comment by Horusiath** [comment]
The case of explicit type annotations is intentional - we don't want to accidentally break the existing API, which is **very easy** to do when you're making changes and let compiler to implicitly infer the types. Since Akka.FSharp is not longer in beta, this is an important factor for us. 

Also some of your ideas concerning code cleanup - like better splitting modules into separate files and using namespace.module declarations to reduce testing - are actually nice.

You may also want to look at some of `fsharp` annotated issues on core akka github [tracker](https://github.com/akkadotnet/akka.net/issues?q=is%3Aopen+is%3Aissue+label%3Afsharp). You may find something interesting back there, or may want to add something from yourself ;)

File: src/core/Akka.FSharp/Spawning.fs [hunk 4]
```diff
@@ -12,129 +12,135 @@ open Microsoft.FSharp.Quotations
...
-    /// Wraps provided function with actor behavior. 
-    /// It will be invoked each time, an actor will receive a message. 
+    /// Wraps provided function with actor behavior.
+    /// It will be invoked each time, an actor will receive a message.
     /// </summary>
-    let actorOf (fn : 'Message -> unit) (mailbox : Actor<'Message>) : Cont<'Message, 'Returned> = 
-        let rec loop() = 
-            actor { 
-                let! msg = mailbox.Receive()
+    let actorOf (fn : 'Message -> unit) : Actor<'Message> -> Cont<'Message, 'Returned> =
```
Comment (Horusiath) [inline src/core/Akka.FSharp/Spawning.fs:126]: This new signature describes an intent better IMO. We could make use of it.

**Comment by neoeinstein** [comment]
Closing, I may submit a cleaned up PR with the `actorOf` signature updates, but thanks for explaining things to me. I look forward to getting a new version of the FSharp API release so I can begin using the typed API.

**Pull request closed without merging**